joomla/joomla-cms
# [4.0] Rework tags to have a dedicated interface

Pull Request for Issue #31127 (and addresses comments made in the original PR https://github.com/joomla/joomla-cms/pull/31793) .

I don't like making b/c breaks in functionality this late into the J4 development. But I think this is required before we start wrapping ourselves in circles with workarounds in batch funtionality.

### Summary of Changes
Reworks some of tags onto an interface (partly stolen from @Hackwar 's work in https://github.com/joomla/joomla-cms/pull/24551 but without some of the larger changes) like we already did with versions.

There's also a new plugin event specifically for batch events that will allow us to to transformations of objects on the data being created. I chose to have a plugin here as the newTag property is kinda internal to the taggable plugin. Whereas the other batch methods can simply use the interface and don't really have any tags logic (other than using for the interface methods)

The new `TaggableTableInterface` should also allow us to cleanup any other interactions (such as the existing saveorder code).

### Testing Instructions
Test tags functionality in all components (articles, categories, contact and newsfeeds). Confirm assigning, removing tags to all these things work. Confirm all still work

Test batch moving items to a new category or language. Before patch tags would get lost, after patch they don't.

Test batch copying items. Before patch tags are not preserved after patch they are.

### Documentation Changes Required
Interface and trait needs to be documented. We also should backport them as empty back to the 3.10 release like we did with the versioning
